text/0477-policy-validation.md

```diff
@@ -0,0 +1,548 @@
+  application.
+* Developer adds the validations property, and avoids these violations going forward.
+
+#### Alternative
+
+```ts
+const app = new App();
+Policy.of(app).add(new CfnGuardValidator());
+```
+
+When you synthesize the CDK app the validator plugins will be called and the results will be printed.
+
+```text
+$ cdk synth
+
+Validation Report (CfnGuardValidator)
+-------------------------------------
+
+(Summary)
+
+╔═══════════╤════════════════════════╗
+║ Status    │ failure                ║
+╟───────────┼────────────────────────╢
+║ Plugin    │ CfnGuardValidator      ║
+╚═══════════╧════════════════════════╝
+
+(Violations)
+
+Ensure S3 Buckets are encrypted with a KMS CMK (2 occurrences)
+
+  Occurrences:
+  
+    - Construct Path: cdk-app/MyStack/Bucket/Resource
+    - Stack Template Path: ./cdk.out/MyStack.template.json
+    - Creation Stack: [
+        'new Bucket (/home/hallcor/tmp/cdk-tmp-app/node_modules/aws-cdk-lib/aws-s3/lib/bucket.js:1:12889)',
+        'new MyStack (/home/hallcor/tmp/cdk-tmp-app/src/main.ts:9:20)',
+      ]
+    - Resource Name: my-bucket
+    - Locations:
+      > BucketEncryption/ServerSideEncryptionConfiguration/0/ServerSideEncryptionByDefault/SSEAlgorithm
+
+  Recommendation: Missing value for key `SSEAlgorithm` - must specify `aws:kms`
+  How to fix:
+    > Using override `app.findChild('my-bucket').addPropertyOverride('SSEAlgorithm', 'aws:kms');`
+    > Add to construct properties for `cdk-app/MyStack/Bucket`
+      `encryption: BucketEncryption.KMS`
+
+  
+Validation failed. See above reports for details
+```
+
+### Adding Exemptions/Suppressions
+
+> Prior art: [cdk-nag](https://github.com/cdklabs/cdk-nag).
+
+There may be cases where you would like to suppress a certain rule. To do so you must specify the rule to suppress and
+the reason for the suppression. Each plugin is responsible for handling it's own
+suppressions and exposes a standard API.
+
+A suppression can be added for all resources under a construct scope. For
+example to add suppressions for an entire stack.
+
+```ts
+import { Stack } from 'aws-cdk-lib';
+import { CfnGuardValidator } from '@aws-cdk/cfn-guard-validator';
+
+const stack = new Stack(app, 'DevStack');
+CfnGuardValidator.addSuppression(stack, {
+  rule: 'S3BucketEncryption'
+  reason: 'Dev environment buckets do not have to be encrypted'
+});
+```
+
+They can also be added for specific resources.
+
+```ts
+import { CfnGuardValidator } from '@aws-cdk/cfn-guard-validator';
+
+class MyStack extends Stack {
+  constructor(scope: Construct, id: string, props?: StackProps) {
+    const bucket = new Bucket(this, 'MyBucket');
+    CfnGuardValidator.addSuppression(bucket, {
+      rule: 'S3BucketEncryption',
+      reason: 'This bucket does not require encryption because xyz',
+    });
+  }
+}
+```
+
+Each plugin will determine how the suppressions are handled for the specific tool used, for example a plugin might
+add to the resource metadata.
```

Comment:
yeah it would then read it during validation. That way you could have a single guard rule, for example

```
let s3_bucket_public_read_prohibited = Resources.*[ Type == 'AWS::S3::Bucket'
  Metadata.guard.SuppressedRules not exists or
  Metadata.guard.SuppressedRules.* != "S3_BUCKET_PUBLIC_READ_PROHIBITED"
]

rule S3_BUCKET_PUBLIC_READ_PROHIBITED when %s3_bucket_public_read_prohibited !empty {
  %s3_bucket_public_read_prohibited.Properties.PublicAccessBlockConfiguration exists
  %s3_bucket_public_read_prohibited.Properties.PublicAccessBlockConfiguration.BlockPublicAcls == true
  %s3_bucket_public_read_prohibited.Properties.PublicAccessBlockConfiguration.BlockPublicPolicy == true
  %s3_bucket_public_read_prohibited.Properties.PublicAccessBlockConfiguration.IgnorePublicAcls == true
  %s3_bucket_public_read_prohibited.Properties.PublicAccessBlockConfiguration.RestrictPublicBuckets == true
  <<
    Violation: S3 Bucket Public Write Access controls need to be restricted.
    Fix: Set S3 Bucket PublicAccessBlockConfiguration properties for BlockPublicAcls, BlockPublicPolicy, IgnorePublicAcls, RestrictPublicBuckets parameters to true.
  >>
}
```

This rule could be evaluated by running guard as part of a guard plugin _or_ it could be run in a CFN Hook when the cloudformation template is actually deployed. The suppression needs to be persisted in a way that allows it to be evaluated during deployment as well, so rather than create a separate suppression mechanism that only works for CDK plugins, we're trying to create one that works for both.

Reply:
What you are describing is worse than "the plugin takes care of it" though. This is "the user-written ruleset must take suppressions into account".

I wonder if these aren't in actuality two different features:

* One to actually suppress findings (I think this might be good to have anyway, like how people would like suppressions for our warnings, which we currently don't offer)
* One to smuggle out-of-band data from the CDK app to the rule set, so that the rule set can use it in its rules

Reply:
> What you are describing is worse than "the plugin takes care of it" though. This is "the user-written ruleset must take suppressions into account".

I agree it's not great, but what are the alternatives? Using the CFN guard example above, the _only_ option for suppressions is to have the suppression somewhere in the template. Guard doesn't have the ability to read data from somewhere else. So if we wrote some suppression logic into the plugin it would allow the developer to get passed local development, but then it would fail at deployment because the deployment system doesn't know about the suppression you added.

The only other alternative I can think of is to also build a feature into CFN hooks around suppressions. CFN hooks would read suppression info from somewhere (not sure it can be given access to anything other than the template though).

Reply:
After talking to Rico I split this section into two sections, 1 for each of suppressions and exemptions. @rix0rrr let me know what you think.